### PR TITLE
Implement `init`; bring synchronization primitives to spec

### PR DESCRIPTION
**Pull request by hausdorff**

### CODE REVIEW

**Comment by hausdorff** [comment]
Marking this as [WIP] as I push up some more speculative stuff for Jessica Yuen.

File: metadata/clusterspec.go [hunk 1]
```diff
@@ -1,15 +1,49 @@
...
 	"github.com/spf13/afero"
 )
 
+const (
+	k8sVersionURLTemplate = "https://raw.githubusercontent.com/kubernetes/kubernetes/%s/api/openapi-spec/swagger.json"
+)
+
+func parseClusterSpec(specFlag string, fs afero.Fs) (ClusterSpec, error) {
+	split := strings.SplitN(specFlag, ":", 2)
+	if len(split) == 0 || len(split) == 1 || split[1] == "" {
```
Comment (jessicayuen) [inline metadata/clusterspec.go]: if len(split) <= 1 || split[1] == "" { ?

File: cmd/init.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+var initCmd = &cobra.Command{
+	Use:   "init <app-name>",
+	Short: "Initialize a ksonnet project",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		flags := cmd.Flags()
+		if len(args) != 1 {
+			return fmt.Errorf("'init' requires a name for the application we're initializing")
+		}
```
Comment (jessicayuen) [inline cmd/init.go]: message might be misleading for `init My Project`; do you want to handle these cases now?

Comment (hausdorff) [reply cmd/init.go]: Resolved.

File: metadata/clusterspec.go (hunk 1 above)
Comment (hausdorff) [reply metadata/clusterspec.go]: Resolved.

File: cmd/root.go [hunk 3]
```diff
@@ -155,176 +152,49 @@ func (f *logFormatter) Format(e *log.Entry) ([]byte, error) {
...
-}
 
-func buildResolver(cmd *cobra.Command) (utils.Resolver, error) {
-	flags := cmd.Flags()
-	resolver, err := flags.GetString(flagResolver)
-	if err != nil {
-		return nil, err
-	}
-	failAction, err := flags.GetString(flagResolvFail)
+	spec.Resolver, err = flags.GetString(flagResolver)
```
Comment (anguslees) [inline cmd/root.go]: ? The resolver stuff has nothing to do with jsonnet (it's for docker tag->digest conversion).  Why are you moving it to the Expander abstraction ?

File: template/expander.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	ExtVars     []string
+	ExtVarFiles []string
+	TlaVars     []string
+	TlaVarFiles []string
+
+	Resolver   string
+	FailAction string
+}
+
+func (spec *Expander) Expand(paths []string) ([]*unstructured.Unstructured, error) {
```
Comment (anguslees) [inline template/expander.go]: This is an awkward library API.  This function now builds it's own file paths, does file accesses, looks up environment variables, etc, which makes it very hard to reuse this code in any other situation (even unittests).  Can we define a narrower scope for what an "Expander" actually does, and refactor some of these environmental dependencies back out of this package?

File: template/resolver.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/ksonnet/kubecfg/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+func (spec *Expander) buildResolver() (utils.Resolver, error) {
```
Comment (anguslees) [inline template/resolver.go]: repeating earlier comment, this doesn't belong here - these are implementation details of the resolveImage() native jsonnet function and should be moved closer to the other native functions.

**Pull request merged**